django/db/models/indexes.py

```diff
@@ -0,0 +1,134 @@
+
+    def create_sql(self, schema_editor, suffix=''):
+        raise NotImplementedError('subclasses of BaseIndex require a create_sql() method')
+
+    def remove_sql(self, schema_editor):
+        raise NotImplementedError('subclasses of BaseIndex require a remove_sql() method')
+
+    def _hash_generator(*args):
+        """
+        Generate a 32-bit digest of a set of arguments that can be used to
+        shorten identifying names.
+        """
+        h = hashlib.md5()
+        for arg in args:
+            h.update(force_bytes(arg))
+        return h.hexdigest()[:8]
+
+    def get_name(self, suffix=None):
+        """
+        Generate a unique name for the index.
+        """
+        if suffix is None:
+            suffix = '_%s' % self.index_type
+        column_names = self.fields
+        # If there is just one column in the index, use a default algorithm from Django
+        table_name = self.model._meta.db_table
+        index_unique_hash = self._hash_generator(column_names)
+        index_name = ('%s_%s_%s%s' % (
+            table_name, column_names[0], index_unique_hash, suffix
+        )).replace('"', '').replace('.', '_')
+        # If the name is too long chop off characters from the prefix_part of the name
+        if len(index_name) > MAX_NAME_LENGTH:
+            # If table name is too long keep it in the prefix_part
+            prefix_part = table_name
+            suffix_part = ('_%s_%s%s' % (column_names[0], index_unique_hash, suffix))
+            if len(suffix_part) > MAX_NAME_LENGTH:  # When column_names[0] is too long
+                # If column name is too long keep it in the prefix_part
+                prefix_part = ('%s_%s' % (table_name, column_names[0]))
+                suffix_part = ('_%s%s' % (index_unique_hash, suffix))
+            index_name = '%s%s' % (
+                prefix_part[:(MAX_NAME_LENGTH - len(suffix_part))], suffix_part
+            )
+        # It shouldn't start with an underscore (Oracle hates this)
+        if index_name[0] == "_":
+            index_name = index_name[1:]
+        # It can't start with a number on Oracle, so prepend D if we need to
+        if index_name[0].isdigit():
+            index_name = "D%s" % index_name[1:]
+        return index_name
+
+    def deconstruct(self):
+        """
+        Returns a 3-tuple of class import path, positional arguments, and keyword
+        arguments.
+        """
+        kwargs = {
+            'type': self.index_type,
+        }
+        return (self.__class__.__name__, self.fields, kwargs)
+
+    def __repr__(self):
+        return '<%s: fields="%s">' % (self.__class__.__name__, ', '.join(self.fields))
+
+    def __eq__(self, other):
+        return (self.__class__ == other.__class__) and (self.deconstruct() == other.deconstruct())
+
+    def __ne__(self, other):
+        return not (self == other)
+
+
+class Index(BaseIndex):
+
+    index_type = 'idx'
+
+    def create_sql(self, schema_editor, suffix=''):
```

Comment:
`create_index_name` is a method of `BaseDatabaseSchemaEditor` because it's schema dependent whereas in `get_name` I have made it independent of schema so that we don't have different names for the same index for users using different schemas (as we want `name` to be a part of the deconstructed index class in the migrations).
About `_hash_generator` it is function having nothing to do with `SchemaEditor` nor with `Index`. It should be made a separate function instead of methods of these two classes IMO.